### PR TITLE
Enhancement/flow expiration

### PR DESCRIPTION
**Pull request by matissecallewaert**
Enhanced the flow termination logic using a priority queue so only the flows that are older are checked instead of the whole map.

### CODE REVIEW

File: rustiflow/src/main.rs [hunk 1]
```diff
@@ -1102,55 +975,78 @@ fn process_packet_ipv6<T>(
...
         }
     });
 
+    let flows_to_remove = {
+        let expirations_to_remove = expirations.lock().unwrap();
+        let mut flows_to_remove = Vec::new();
+
+        for (flow_id, &expiration) in expirations_to_remove.iter() {
+            if expiration <= ts {
+                flows_to_remove.push(flow_id.clone());
```
Comment (mielverkerken) [inline rustiflow/src/main.rs:1021]: I believe exporting the expired flows is missing:
```
if *NO_CONTAMINANT_FEATURES.lock().unwrap().deref() {
      export(&flow.dump_without_contamination());
} else {
      export(&flow.dump());
}
```

**Review by mielverkerken — CHANGES_REQUESTED** [review]

File: rustiflow/src/main.rs [hunk 2]
```diff
@@ -1102,55 +975,78 @@ fn process_packet_ipv6<T>(
...
+            } else {
+                break;
+            }
+        }
+
+        flows_to_remove
+    };
```
Comment (mielverkerken) [inline rustiflow/src/main.rs:1028]:
```
let flows_to_remove = expirations.lock().unwrap()
        .iter()
        .take_while(|&(_, &expiration)| expiration <= ts)
        .map(|(flow_id, _)| flow_id.clone())
        .collect::<Vec<String>>();
```

File: rustiflow/src/main.rs [hunk 3]
```diff
@@ -1102,55 +975,78 @@ fn process_packet_ipv6<T>(
...
+
+    // Remove the expired flows
+    {
+        let mut expirations = expirations.lock().unwrap();
+
+        for flow_id in flows_to_remove {
+            expirations.remove(&flow_id);
+            flow_map.remove(&flow_id);
+        }
+    }
```
Comment (mielverkerken) [inline rustiflow/src/main.rs:1038]: Wrapping block needed?

File: rustiflow/src/main.rs [hunk 4]
```diff
@@ -1102,55 +975,78 @@ fn process_packet_ipv6<T>(
...
-        export(&end.unwrap());
+        let flow = entry.value();
+        if *NO_CONTAMINANT_FEATURES.lock().unwrap().deref() {
+            export(&flow.dump_without_contamination());
+        } else {
+            export(&flow.dump());
+        }
         drop(entry);
-        flow_map.remove(&flow_id_remove);
+        flow_map.remove(&flow_id);
```
Comment (mielverkerken) [inline rustiflow/src/main.rs:1049]:
```
    if entry.update_flow(&features, &ts, ts_date, fwd).is_some() {
        let flow = entry.value();
        let dump = if *NO_CONTAMINANT_FEATURES.lock().unwrap() {
            flow.dump_without_contamination()
        } else {
            flow.dump()
        };
        export(&dump);
        drop(entry);
        flow_map.remove(&flow_id);
    }

**Review by Str-Gen — APPROVED** [review]

File: rustiflow/src/main.rs (hunk 3 above)
Comment (matissecallewaert) [reply rustiflow/src/main.rs:1038]: yes, otherwise I either had deadlocks or some errors about scopes.

File: rustiflow/src/main.rs (hunk 4 above)
Comment (matissecallewaert) [reply rustiflow/src/main.rs:1049]: Out of the entry.update_flow you already get the output string, what you suggest is just not how the function works.

Comment (mielverkerken) [reply rustiflow/src/main.rs:1049]: In the code, the output of `update_flow` was also unused. We can reduce code even further:
```
    if let Some(flow) = entry.update_flow(&features, &ts, ts_date, fwd) {
        export(&flow);
        drop(entry);
        flow_map.remove(&flow_id);
    }
```

Comment (matissecallewaert) [reply rustiflow/src/main.rs:1049]: My bad, was indeed incorrect, the update_flow already gives either the contaminant output or the normal one.

**Comment by matissecallewaert** [comment]
I still have a deadlock on line 988 that I can't seem to fix, it also does not happen always that that line gives a deadlock.

**Review by Str-Gen — APPROVED** [review]

**Comment by matissecallewaert** [comment]
@mielverkerken is this PR still relevant, because I think the major refactor branch already addresses this issue?

**Pull request closed without merging**